src/contracts/permissions/PermissionController.sol

```diff
@@ -0,0 +1,176 @@
+        _;
+    }
+
+    /**
+     *
+     *                         INITIALIZING FUNCTIONS
+     *
+     */
+    constructor() {
+        _disableInitializers();
+    }
+
+    function initialize() external initializer {}
+
+    /**
+     *
+     *                         EXTERNAL FUNCTIONS
+     *
+     */
+
+    /// @inheritdoc IPermissionController
+    function setAdmin(address account, address admin) external onlyAdmin(account) {
+        // Add the admin to the account's admins
+        // If the admin is already set, the set will fail
+        EnumerableSet.AddressSet storage admins = _permissions[account].admins;
+        require(admins.add(admin), AdminAlreadySet());
+
+        emit AdminSet(account, admin);
+    }
+
+    function removeAdmin(address account, address admin) external onlyAdmin(account) {
+        EnumerableSet.AddressSet storage admins = _permissions[account].admins;
+
+        // Remove the admin from the account's admins
+        // If the admin is not set, the remove will fail
+        require(admins.remove(admin), AdminNotSet());
+
+        emit AdminRemoved(account, admin);
+    }
+
+    /// @inheritdoc IPermissionController
+    function setDelegate(
+        address account,
+        address delegate,
+        address target,
+        bytes4 selector
+    ) external onlyAdmin(account) {
+        AccountPermissions storage permissions = _permissions[account];
+
+        bytes32 targetSelector = _encodeTargetSelector(target, selector);
+        require(!permissions.delegatePermissions[delegate].contains(targetSelector), DelegateAlreadySet());
+
+        // Add the delegate to the account's permissions
+        permissions.delegatePermissions[delegate].add(targetSelector);
+        permissions.permissionDelegates[targetSelector].add(delegate);
+
+        emit DelegateSet(account, delegate, target, selector);
+    }
+
+    /// @inheritdoc IPermissionController
+    function removeDelegate(
+        address account,
+        address delegate,
+        address target,
+        bytes4 selector
+    ) external onlyAdmin(account) {
+        AccountPermissions storage permissions = _permissions[account];
+
+        bytes32 targetSelector = _encodeTargetSelector(target, selector);
+        require(permissions.delegatePermissions[delegate].contains(targetSelector), DelegateNotSet());
+
+        // Remove the delegate from the account's permissions
+        permissions.delegatePermissions[delegate].remove(targetSelector);
+        permissions.permissionDelegates[targetSelector].remove(delegate);
+
+        emit DelegateRemoved(account, delegate, target, selector);
+    }
+
+    /**
+     *
+     *                         INTERNAL FUNCTIONS
+     *
+     */
+
+    /// @dev Encodes the target and selector into a single bytes32 value
+    function _encodeTargetSelector(address target, bytes4 selector) internal pure returns (bytes32) {
+        return bytes32(abi.encodePacked(target, uint96(bytes12((selector)))));
```

Comment:
Pretty sure this could be simplified to
```
bytes32(abi.encodePacked(target, selector));
```
Decoding would be 

```
address target = address(uint160(uint256(targetSelector) >> 32));
bytes4 selector = bytes4(uint32(uint256(targetSelector));
```

or 
```
(address target, bytes4 selector) = abi.decode(bytes(targetSelector), (address, bytes4));
```

Reply:
without casting in the decode
```
(address target, bytes4 selector) = abi.decode(targetSelector, (address, bytes4));
```

Reply:
Don't think any of those work??

`bytes32(abi.encodePacked(target, selector))`

The encoded would be: [160 bits target][32 bits selector][64 bits padding]

`bytes4 selector = bytes4(uint32(uint256(targetSelector));`

Ends up being all zeros because we're getting the least significant 32 bits of the encoded value. 

Can't do abi.decode since it depends on bytes and would rather not do the bytes.concat(bytes32).